File: lib/guillotine/parser/quotes_parser.c
#include <stdbool.h>
#include <string.h>
#include <stdio.h>

#define SINGLE_QUOTE '\''
#define DOUBLE_QUOTE '\"'
#define BACKTICK     '`'
#define SPACE        ' '

char   original_string[1000];
char   buffer[1000] = "";
int    at        = 0;
bool   in_quotes = false;
char   starting_quote;

char * parse_and_eval();
bool chars_left();
char current_char();
char next_char();
void eat_whitespace();
void add_current_char();
void add_to_buffer(char);
void update_quote_status();
int advance_char();
bool a_quote(char);

int main(int argc, char **argv) {
  strcpy(original_string, argv[1]);
  parse_and_eval();
  printf("%s", buffer);
  return 0;
}

char * parse_and_eval() {
  while (chars_left()) {
    eat_whitespace();

    if (chars_left()) {
      add_current_char();
    }
  }
  return buffer;
}

bool chars_left() {
  return (at < strlen(original_string));
}

char current_char() {
  return original_string[at];
}

char next_char() {
  return original_string[at + 1];
}

void eat_whitespace() {
  if (in_quotes == false && current_char() == SPACE && next_char() == SPACE) {
    advance_char();
    eat_whitespace();
  }
}

void add_current_char() {
  update_quote_status();
  add_to_buffer(current_char());
  advance_char();
}

int advance_char() {
  return at++;
}

void add_to_buffer(char c) {
  char array[2] = "";
  array[0] = c;
  strcat(buffer, array);
}

void update_quote_status() {
  char c = current_char();

  if (in_quotes == true) {
    if (c == starting_quote) {
      in_quotes = false;
      starting_quote = c;
    }
  } else if (a_quote(c)) {
    in_quotes = true;
    starting_quote = c;
  }
}

bool a_quote(char c) {
  return(c == SINGLE_QUOTE || c == DOUBLE_QUOTE || c == BACKTICK);
}

File: lib/guillotine/parser/quotes_parser.rb
module Guillotine
  module Parser
    class QuotesParser
      SINGLE_QUOTE = "'"
      DOUBLE_QUOTE = '"'
      BACKTICK     = '`'
      SPACE = " "
      
      def initialize
        @at = 0
        @buffer = ""
        @in_quotes = false
      end
      
      def parse(string)
        @original_content = string.strip
        @length = @original_content.length

        process_char while chars_left?
        
        self
      end
      
      def eval
        @buffer
      end
      
    private
      
      def process_char
        eat_whitespace
        add_current_char if chars_left?
      end
      
      def add_current_char
        update_quote_status
        @buffer << current_char
        advance_char
      end
      
      def update_quote_status
        if @in_quotes
          if current_char == @start_quote
            @in_quotes = false
            @start_quote = nil
          end
        elsif current_char_a_quote?
          @in_quotes = true
          @start_quote = current_char
        end
      end
      
      def current_char_a_quote?
        current_char == SINGLE_QUOTE || current_char == DOUBLE_QUOTE || current_char == BACKTICK
      end
      
      def advance_char
        @at += 1
      end
      
      def eat_whitespace
        if @in_quotes == false && current_char == SPACE && next_char == SPACE
          advance_char 
          eat_whitespace
        end
      end
      
      def current_char
        @original_content[@at].chr
      end
      
      def next_char
        if next_char = @original_content[@at+1]
          next_char.chr
        end
      end
      
      def chars_left?
        @length > @at
      end
    end
  end
end
